lib/web/apiserver.go

```diff
@@ -1294,13 +1309,45 @@ func (h *Handler) logout(w http.ResponseWriter, ctx *SessionContext) error {
 	return nil
 }
 
-// renewSession is called in two ways:
-// 	- Without requestId: Updates the existing session about to expire.
-// 	- With requestId: Updates existing session with additional roles assigned with approving access request.
+type renewSessionRequest struct {
+	// AccessRequestID is the id of an approved access request.
+	AccessRequestID string `json:"requestId"`
+	// Expires defines when the new session should expire.
+	Expires time.Time `json:"expires"`
+	// Roles is the list of roles that the new session should be assigned.
+	Roles []string `json:"roles"`
+}
+
+// renewSession updates this existing session with a new session and sets cookie with the new session.
+//
+// Depending on request fields sent in with request, the new session creation can be modified:
+//   - requestId (opt): appends roles approved from access request to currently assigned roles or,
+//   - expires and roles (opt): values that were preserved from first session creation after web login.
+//     Allows users who assumed roles by approved access requests, to switch back to their default roles
+//     and this new session lives through remainining expiration time.
 func (h *Handler) renewSession(w http.ResponseWriter, r *http.Request, params httprouter.Params, ctx *SessionContext) (interface{}, error) {
-	requestID := params.ByName("requestId")
+	req := &renewSessionRequest{}
+	if r.Body != http.NoBody {
+		if err := httplib.ReadJSON(r, &req); err != nil {
```

Comment:
nit: this should not use the address of the pointer, I don't see a reason for indirection. Just use a local value of type `renewSessionRequest`:
```go
var req renewSessionRequest
```